src/audio/google/google_rtc_audio_processing.c

```diff
@@ -543,223 +556,324 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
+	int ref_chan = source_get_channels(sources[cd->aec_reference_source]);
+	int ref_rate = source_get_rate(sources[cd->aec_reference_source]);
+
+	int mic_fmt = source_get_frm_fmt(sources[cd->raw_microphone_source]);
+	int mic_chan = source_get_channels(sources[cd->raw_microphone_source]);
+	int mic_rate = source_get_rate(sources[cd->raw_microphone_source]);
+
+	int out_fmt = sink_get_frm_fmt(sinks[0]);
+	int out_chan = sink_get_channels(sinks[0]);
+	int out_rate = sink_get_rate(sinks[0]);
+
+	/* Too many channels is a soft failure, AEC treats only the first N */
+	if (ref_chan > REF_CHAN_MAX)
+		comp_warn(dev, "Too many ref channels: %d, truncating to %d",
+			  ref_chan, REF_CHAN_MAX);
+	if (mic_chan > MIC_CHAN_MAX)
+		comp_warn(dev, "Too many mic channels: %d, truncating to %d",
+			  mic_chan, MIC_CHAN_MAX);
+
+	if (out_chan != mic_chan) {
+		comp_err(dev, "Input/output mic channel mismatch");
+		ret = -EINVAL;
 	}
 
-	audio_stream_init_alignment_constants(1, 1, &output->stream);
-	frame_fmt = audio_stream_get_frm_fmt(&output->stream);
-	rate = audio_stream_get_rate(&output->stream);
-	output_stream_channels = audio_stream_get_channels(&output->stream);
+	cd->num_aec_reference_channels = MIN(ref_chan, REF_CHAN_MAX);
+	cd->num_capture_channels = MIN(mic_chan, MIC_CHAN_MAX);
 
-	if (cd->num_capture_channels > microphone_stream_channels) {
-		comp_err(dev, "unsupported number of microphone channels: %d",
-			 microphone_stream_channels);
-		return -EINVAL;
+	if (ref_rate != mic_rate || ref_rate != out_rate ||
+	    ref_rate != CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ) {
+		comp_err(dev, "Incorrect source/sink sample rate, expect %d\n",
+			 CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ);
+		ret = -EINVAL;
 	}
 
-	if (cd->num_capture_channels > output_stream_channels) {
-		comp_err(dev, "unsupported number of output channels: %d",
-			 output_stream_channels);
-		return -EINVAL;
+	if (ref_fmt != bits_fmt(CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_REF_BITS)) {
+		comp_err(dev, "Reference stream must be %d bit samples\n",
+			 CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_REF_BITS);
+		ret = -EINVAL;
 	}
 
-	switch (frame_fmt) {
-#if CONFIG_FORMAT_S16LE
-	case SOF_IPC_FRAME_S16_LE:
-		break;
-#endif /* CONFIG_FORMAT_S16LE */
-	default:
-		comp_err(dev, "unsupported data format: %d", frame_fmt);
-		return -EINVAL;
+	if (mic_fmt != bits_fmt(CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_MIC_BITS)) {
+		comp_err(dev, "Mic stream must be %d bit samples\n",
+			 CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_MIC_BITS);
+
+		// FIXME: squash for now, the streams lie.  See below.
+		//ret = -EINVAL;
 	}
 
-	if (rate != CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ) {
-		comp_err(dev, "unsupported samplerate: %d", rate);
-		return -EINVAL;
+	if (mic_fmt != out_fmt) {
+		comp_err(dev, "Mismatched in/out frame format");
+		ret = -EINVAL;
 	}
 
+	// FIXME: the streams have a bad format on MTL, so we can't
```

Comment:
Vaguely.  It seems like a design bug, but I haven't had time to chase things down or submit a proper bug report.   So, recognizing that this is just my understanding and that some of it is wrong:

With IPC3, the comp_buffer objects are explicitly represented in the topology, so they get sample formats set from there at creation time.  With IPC4, they're created implicitly as part of the "edges" between widgets.  So they would have to get the value dynamically from the components they are connected to.

But you can't ask a component what it's sample format is, because that data has always been stored inside the audio_stream in the buffer!  So it's stuck in a chicken/egg situation.  I think the "correct" solution would be to follow the pipeline upstream until you find a buffer or DAI which knows the format, but there's no metadata to distinguish "I don't know my format" from "S16_LE" (which is a numeric zero from initialization).

Presumably someone here has  better idea how to fix this than I do.